Validation check and managing users API

newoo opened the pull request:
과제를 늦은 시간에 시작하여 많은 양을 하지는 못했습니다.

### 금일 진행 사항
- mapper를 활용하여 DTO와 Domain Model를 연결하는 것
- User model 추가 및 빌더를 이용한 생성 테스트

### 의존성 추가사항
`implementation 'com.github.dozermapper:dozer-core:6.4.0'`

DTO와 도메인 모델을 맵핑하기 위해 추가되었습니다.
참고 [github](https://github.com/CodeSoom/spring-week5-assignment-1/pull/5#issue-577581640)

newoo committed:
Apply product mapping in createProduct method

newoo committed:
Apply product mapping at updateProduct method

newoo committed:
Create user model

newoo committed:
Add and pass user creation test with builder

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -0,0 +1,22 @@
+        assertThat(user.getEmail()).isEqualTo("newoo@codesoom.com");
+        assertThat(user.getPassword()).isEqualTo("codesoom123");
+    }
+}
```
파일 마지막에 개행 문자를 추가해 주세요.

johngrib commented on app/build.gradle:
```diff
@@ -51,6 +51,9 @@ dependencies {
     annotationProcessor 'org.projectlombok:lombok:1.18.16'
 
+    // DozerMapper
+    implementation 'com.github.dozermapper:dozer-core:6.4.0'
```
dozermapper를 의존성에 추가하셨군요! 이렇게 새로운 라이브러리를 `build.gradle`에 추가할 때에는 팀 동료를 위해 PR에 나름의 설명을 추가해주시면 좋겠죠? 이른바 "자기 PR에 자신이 리뷰를 남겨서 리뷰어의 이해를 돕기 위한 설명을 추가하는" 방법입니다.

PR에 라이브러리의 공식 문서의 링크 등을 추가해 주시고, 간단히 기능을 소개하고, 이런이런 장점이 있어서 도입을 고려하였다고 써주시면 좋은 연습이 될 거에요.

newoo replied:
가장 상단 코멘트에 관련사항을 추가하여 업데이트 해보았습니다.

johngrib reviewed (commented):
이번 한 주도 힘내서 화이팅!

newoo committed:
Add test about changeWith method of User model

newoo committed:
Pass test about changeWith method

Complete changeWith method to udpate name, email, password of user model

newoo committed:
Delete NoArgsConstructor annotation for code coverage

newoo committed:
Add toString method test of user model

newoo committed:
Pass toString method test of user model and builder

newoo committed:
Add toString method test of product model

newoo committed:
Pass toString method of product model and builder

Override toString method of product model to pass test

newoo committed:
Add DTO for user model

newoo committed:
Create UserService and add getProducts method test

newoo committed:
Create UserRepository

newoo committed:
Pass getUsers method tests

newoo committed:
Add createUser method test

newoo committed:
Pass createUser method test

To pass test, add annotations and update codes.

newoo commented:
코드 커버리지 100%보다는 기본적인 기능 테스트가 우선 진행되었어야 했는데, 그렇게 하지 못했다고 생각합니다.(우선순위 정렬 실패)
내일은 요구사항에 대한 기능을 추가하는 것을 최우선으로 진행하려고 합니다.

### 금일 진행 사항

- 도메인 관련 테스트 코드 커버리지 100% 목표로 진행
- UserService 관련 테스트 코드 작성 및 테스트 통과 진행
  > User 리스트를 리턴하는 메서드와 User를 추가하는 메서드에 대한 테스트

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -27,24 +30,14 @@ public Product getProduct(Long id) {
-                productData.getPrice(),
-                productData.getImageUrl()
-        );
+        product.changeWith(mapper.map(productData, Product.class));
```
앗 뭔가 추가로 해야 할 일 하나가 빠지지는 않았나요?

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,31 @@
+
+@Service
+@Transactional
+public class UserService {
```
함께 작업하는 동료를 위해 서비스와 컨트롤러에는 모두 JavaDoc을 작성하는 습관을 들이는 것이 좋습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Product.java:
```diff
@@ -27,13 +27,15 @@
+
+    @Override
+    public String toString() {
+        return String.format("%s 장난감", name);
```
`toString` 구현 내용이 조금 더 자세하면 좋겠어요. 기왕 lombok을 쓰는 김에 lombok의 도움을 좀 받는 것도 좋겠죠.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,113 @@
+        class Context_without_any_saved_user {
+            @BeforeEach
+            void setEmptyList() {
+                given(userRepository.findAll()).willReturn(List.of());
```
`.willReturn`에서 한 번 개행해주면 좋을 것 같아요.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,113 @@
+        @Test
+        @DisplayName("user를 추가하고, 추가된 user를 리턴한다.")
+        void it_create_user_and_return_created_user() {
+            given(userRepository.save(any(User.class))).will(invocation -> {
+                return invocation.getArgument(0);
+            });
```
이 부분은 테스트를 준비하는 코드입니다. `context`로 분류해 주시면 좋겠어요.

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/ProductTest.java:
```diff
@@ -23,20 +29,23 @@ void creationWithBuilder() {
+    @Test
+    void toStringTest() {
+        assertThat(product.toString()).isEqualTo("%s 장난감", product.getName());
+        assertThat(Product.builder().toString()).isEqualTo("Product.ProductBuilder(id=null, name=null, maker=null, price=null, imageUrl=null)");
```
한 줄이 너무 깁니다. 적절히 개행해 주시면 좋겠어요.

newoo committed:
Add getUser method of UserService tests

newoo committed:
Update getUser method

newoo committed:
Create UserNotFoundException and pass getUser method tests

newoo committed:
Add updateUser and deleteUser signature

newoo committed:
Add updateUser method test

newoo committed:
Update updateUser method and pass test

newoo committed:
Add toString annotation at models

newoo commented:
아직 PR 피드백에 대한 수정을 진행하지는 못했습니다.
과제진행 중이나 PR 먼저 업데이트합니다

### 금일 진행 사항

- UserService의 getUser, updateUser 메소드 테스트 코드 작성 및 테스트 통과 진행

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -24,6 +24,10 @@ public UserService(Mapper dozerMapper, UserRepository userRepository) {
 
+    public User getUser(Long id) {
+        return null;
+    }
```
이건 실패하는 테스트를 작성하기 위해 먼저 만든 메서드인가요?

newoo replied:
처음에 테스트 작성하기 전에 메소드 시그니처만 먼저 만들어놓기위해 거짓 리턴값을 넣어두었는데, 안 좋은 습관일까요?

johngrib replied:
테스트코드를 제때 만들어놓아서 테스트 코드가 실패한다면 전혀 나쁜 습관이 아닙니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -88,6 +97,53 @@ void it_return_user_list() {
         }
     }
 
+    @Nested
+    @DisplayName("getUser 메소드는")
+    class Describe_getUser {
+        private Long givenId;
+
+        @Nested
+        @DisplayName("저장된 user의 id를 가지고 있다면")
+        class Context_with_saved_id {
+            private User savedUser;
+
+            @BeforeEach
+            void setSavedId() {
+                givenId = givenSavedId;
+
+                given(userRepository.findById(givenId)).willReturn(Optional.of(user));
+            }
+
+            @Test
+            @DisplayName("user를 리턴한다.")
+            void it_return_user() {
+                savedUser = userService.getUser(givenId);
+
+                verify(userRepository).findById(givenId);
+
+                assertCreatedUser(savedUser);
+            }
+        }
```
`getUser` 메소드를 주제로 삼는 `Describe`이므로, 이 실험의 주제가 되는 부분을 `subject`로 분리해 봅시다.

```java
    @Nested
    @DisplayName("getUser 메소드는")
    class Describe_getUser {
        User subject(String givenId) {
            return userService.getUser(givenId);
        }
```

이렇게 `getUser` 메소드를 `subject`로 분리하면 (마치 실험 보고서처럼) 실험의 주제를 스코프 상단에 표현할 수 있고, 주제를 `subject` 메소드 내에 격리하는 모양이 되어, `getUser` 메소드의 이름이 바뀐다 해도 `subject` 안쪽에서만 변경이 일어나게 됩니다. 테스트 context가 수십개가 되는 상황에서 `getUser` 메소드의 이름을 `findUser`로 변경한다고 상상해 보세요. IDE가 다 바꿔주기는 하겠지만 변경사항은 git에 영원히 남겠죠.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -25,7 +25,7 @@ public UserService(Mapper dozerMapper, UserRepository userRepository) {
 
     public User getUser(Long id) {
-        return null;
+        return userRepository.findById(id).orElseThrow(() -> new UserNotFoundException(id));
```
한 줄이 너무 깁니다. 적절히 개행해 주면 좋겠어요.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -137,7 +137,7 @@ void setUnsavedId() {
             @DisplayName("toy를 찾을 수 없다는 exception을 던진다.")
             void it_throw_exception() {
                 assertThatThrownBy(
-                        () -> UserService.getUser(givenId),
```
이런 오타가 있었군요!

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -164,4 +168,75 @@ void it_create_user_and_return_created_user() {
+                final Mapper mapper = DozerBeanMapperBuilder.buildDefault();
+                modifying = mapper.map(modifyingUserData, User.class);
+
+                given(userRepository.findById(givenId)).willReturn(Optional.of(modifying));
```
만약 `given`을 사용하지 않아야 한다면 어떤 방법으로 테스트하는 것이 좋을지도 생각해 보면 좋겠습니다.

newoo replied:
해당 코드를 예로 들면 `userRepository`를 상속받는 `mockUserRepository`를 추가해서 `findById`를 오버라이드해서 특정값을 리턴할 수 있도록 만들지 않을까 생각합니다. @johngrib 

실제로 iOS에서 테스트할 때 종종 사용하는 방법이기는한데, 적절한 방법일 지는 잘 모르겠습니다.

johngrib replied:
그렇게 할 수도 있을 거고, `@SpringBootTest`를 사용하는 방법도 있을 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Product.java:
```diff
@@ -1,9 +1,6 @@
-import lombok.Builder;
-import lombok.Getter;
-import lombok.NoArgsConstructor;
+import lombok.*;
```
`import *`의 사용을 지양합시다.

johngrib reviewed (commented):
한동안 커밋 메시지를 한국어로 작성해보시면 어떨까요? 커밋 메시지를 섬세하게 작성하는 연습을 한 다음, 영어로 작성하는 방식으로 돌아가도 좋은 학습이 될 거에요.

newoo committed:
import * 사용 제거 및 오타 수정

newoo committed:
긴 코드를 개행처리

newoo committed:
getUsers 테스트 코드에서 getUsers를 suject로 분리

newoo committed:
getUser테스트에서 getUser메서도를 subject로 분리

newoo committed:
createUser테스트에서 createUser메소드를 subject로 분리

newoo committed:
updateUser테스트에서 updateUser메서드를 subject로 분리

newoo committed:
toString메소드관련 테스트 수정 및 긴 코드 개행처리

newoo committed:
updateUser메서드 테스트 코드에서 중복 코드 통합

newoo committed:
createUser메소드 테스트에 setGivenUser메소드를 추가

setGivenUser메소드를 이용해서 userRepository의 save메소드의 리턴값을 처리

newoo committed:
UserService 테스트 코드 내, 긴 코드를 개행처리

newoo committed:
Service 클래스들에 간단한 Javadoc 추가

newoo committed:
ProductService의 메서드들에 JavaDoc추가

newoo committed:
UserService의 deleteUser의 시그니처 변경

newoo committed:
deleteUser메서드 테스트 코드 추가

newoo committed:
deleteUser 코드 내부 완성 및 테스트통과

newoo committed:
UserService의 메소드들의 JavaDoc 추가

newoo commented:
### 금일 진행사항

- deleteUser메소드에 대한 테스트 추가
- service클래스에 JavaDoc추가
- subject를 이용해 테스트코드에서 메서드 분리

johngrib reviewed (commented):
45a256c0490c97dad3b2ed2dd12fc1d7513d11e3 커밋 메시지에 오타가 있습니다. 

`getUser테스트에서 getUser메서도를 subject로 분리` 만약 할 수 있다면 커밋 메시지를 편집해 주시면 좋겠습니다. 하지만 그렇지 않다면 과제의 주제가 git 이 아니므로 넘어가도 좋습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -24,20 +24,43 @@ public UserService(Mapper dozerMapper, UserRepository userRepository) {
+    /**
+     * 유저 리스트를 반환한다.
+     *
+     * @return List<User>, 저장된 유저 리스트
```
콤마는 필요 없습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,315 @@
+            return userService.createUser(userData);
+        }
+
+        @BeforeEach
+        void setGivenUser() {
+            given(userRepository.save(any(User.class)))
+                    .will(invocation -> {
+                        return invocation.getArgument(0);
+                    });
+        }
```
이렇게 고칠 수 있습니다.
```suggestion
        @BeforeEach
        void setGivenUser() {
            given(userRepository.save(any(User.class)))
                    .will(invocation -> invocation.getArgument(0));
        }
```

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -80,7 +80,7 @@ void list() throws Exception {
+    void detailWithExistedProduct() throws Exception {
         mockMvc.perform(
                 get("/products/1")
                         .accept(MediaType.APPLICATION_JSON_UTF8)
```
`MediaType.APPLICATION_JSON_UTF8`은 deprecated 입니다. 다른 것으로 바꿔 주세요. JavaDoc을 잘 읽으면 무엇으로 바꿔야 하는지도 알 수 있습니다.

![image](https://user-images.githubusercontent.com/1855714/109299127-f4e4aa80-7877-11eb-9f05-23a6d1d0b033.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,7 +1,12 @@
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Bean;
+
+import java.beans.BeanProperty;
```
사용하지 않는 import 입니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/UserData.java:
```diff
@@ -0,0 +1,28 @@
+import com.github.dozermapper.core.Mapping;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
```
사용하지 않는 import 입니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/UserRepository.java:
```diff
@@ -0,0 +1,16 @@
+package com.codesoom.assignment.domain;
+
+import org.springframework.data.repository.CrudRepository;
```
사용하지 않는 import 입니다.

newoo committed:
불필요한 import 제거

newoo committed:
APPLICATION_JSON_UTF8를 삭제하고, CharacterEncodingFilter를 이용하여 UTF-8인코딩 처리

newoo committed:
UserController을 생성하고 요구사항에 따라 필요한 메서드의 시그니처 추가

newoo committed:
createUser메서드 테스트 코드 수정

newoo commented:
아직 전체테스트 통과는 못했지만, 남은 시간동안 통과하도록 진행하겠습니다.

johngrib reviewed (commented):
169c370a29137ab2f562e80157c4b869c223a566
커밋을 할 때 작업과 관련된 (예를 들어 리뷰라던가) url을 커밋 본문에 써주면 github에서 자동으로 링크를 걸어줍니다. 이 커밋의 본문엔 이 링크( https://github.com/CodeSoom/spring-week5-assignment-1/pull/5#discussion_r583598289 )가 들어갔다면 적절했겠죠?

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    /**
+     * 유저 정보 수정요청을 받고, 요쳥에 대한 결과를 반환한다.
```
오타가 있습니다. `요쳥` -> `요청`

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * 유저 정보 수정요청을 받고, 요쳥에 대한 결과를 반환한다.
+     *
+     * @param id, 수정할 유저의 id
```
콤마는 제거해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -0,0 +1,54 @@
+ */
+@RestController
+@RequestMapping("/user")
+public class UserController {
```
이 클래스의 JavaDoc이 매우 괜찮게 잘 작성됐습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/UserController.java:
```diff
@@ -0,0 +1,54 @@
+     * 유저 정보 삭제요청을 받고, 요쳥에 대한 결과를 반환한다.
+     *
+     * @param id 삭제할 유저의 id
+     * @return User 삭제된 유저정보
+     */
+    @DeleteMapping("{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void destroy(@PathVariable Long id) {
```
이 메소드는 리턴형이 `void` 이므로 `@return` 태그를 작성할 필요가 없습니다.

newoo committed:
Create UserController test

newoo committed:
POST /users에 대한 응답 테스트 추가

newoo committed:
create메소드 완성 그리고 테스트 실패

newoo committed:
오타수정 그리고 create메소드 테스트 통과

newoo committed:
PATCH /users/{id}에 대한 응답 테스트 2개 추가

newoo committed:
update메소드에 대한 테스트 1개 통과

newoo committed:
update메소드에 대한 전체테스트 통과

newoo committed:
DELETE /users/{id}에 대한 테스트 작성

newoo committed:
destroy메소드 완성 및 테스트 통과

newoo committed:
API 테스트 통과

user모델의 email이 수정할 때는 NotBlank에 해당하지 않기에
CreatingUserData와 UpdatingUserData로 나누어 처리함

newoo committed:
ProductController에 CrossOrigin 어노테이션 추가

newoo committed:
destroy메소드의 JavaDoc 수정

리턴형이 void이지만, JavaDoc에 잘못작성되어 있었음

newoo committed:
create메소드에 대한 JavaDoc 수정

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -101,11 +103,91 @@ void setRequest() {
+
+        @Nested
+        @DisplayName("저장된 user의 id를 가지고 있다면")
+        class Context_with_saved_user_id {
```
문장 구성이 어색합니다. 

- `PATCH /users/{id} 요청에 대한 응답은 - 저장된 user의 id를 가지고 있다면`

더 좋게 고칠 수 있을 것 같아요.

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/UpdatingUserData.java:
```diff
@@ -0,0 +1,26 @@
+@Setter
+@Getter
+@Builder
+public class UpdatingUserData {
```
클래스 이름이 미묘하게 느껴집니다. JavaDoc으로 간결하게 설명해 주세요.

johngrib reviewed (approved):
한 주간 고생 많으셨습니다.
